crates/but-workspace/src/lib.rs

```diff
@@ -43,6 +59,196 @@ pub fn stacks(gb_dir: &Path) -> Result<Vec<StackEntry>> {
         .collect())
 }
 
+/// Represents the state a commit could be in for the purposes of rendering the UI.
+#[derive(Debug, Clone)]
+pub enum CommitState {
+    /// The commit is only local
+    LocalOnly,
+    /// The commit is also present at the remote [`StackBranch::upstream_refrence`].
+    /// This is the commit state if:
+    ///  - The commit has been pushed to the remote
+    ///  - The commit has been copied from a remote commit (when applying a remote branch)
+    ///
+    /// This variant carries the remote commit id.
+    /// The `remote_commit_id` may be the same as the `id` or it may be different if the local commit has been rebased or updated in another way.
+    LocalAndRemote(gix::ObjectId),
+    /// The commit is considered integrated.
+    /// This should happen when this commit or the contents of this commit is already part of the base.
+    Integrated,
+}
+
+/// Commit that is a part of a [`StackBranch`] and, as such, containing state derived in relation to the specific branch.
+#[derive(Debug, Clone)]
+pub struct Commit {
+    /// The OID of the commit.
+    pub id: gix::ObjectId,
+    /// The message of the commit.
+    pub message: BString,
+    /// Whether the commit is in a conflicted state.
+    /// Conflicted state of a commit is a GitButler concept.
+    /// GitButler will perform rebasing/reordering etc without interruptions and flag commits as conflicted if needed.
+    /// Conflicts are resolved via the Edit Mode mechanism.
+    pub has_conflicts: bool,
+    /// Represents wether the the commit is considered integrated, local only,
+    /// or local and remote with respect to the branch it belongs to.
+    /// Note that remote only commits in the context of a branch are expressed with the [`UpstreamCommit`] struct instead of this.
+    pub state: CommitState,
```

Comment:
I wonder if this should be computed lazily (as it's expensive).

Reply:
The current CommitState is not expensive to compute (it's only trivial operations)